V10 Library fails to compile

With github source
[https://github.com/arduino-libraries/WiFi101/tree/9384c5b7b3c39d756877f728fbc346b38871ee77](url)

I get the following compilation errors:

```
Arduino: 1.6.11 (Windows XP), Board: "Arduino/Genuino Mega or Mega 2560, ATmega2560 (Mega 2560)"

C:\Documents and Settings\Gabbitus\My Documents\Arduino\libraries\WiFi101\src\WiFi.cpp: In member function 'uint32_t WiFiClass::getTime()':

C:\Documents and Settings\Gabbitus\My Documents\Arduino\libraries\WiFi101\src\WiFi.cpp:922:13: error: aggregate 'WiFiClass::getTime()::tm tm' has incomplete type and cannot be defined

   struct tm tm;

             ^

C:\Documents and Settings\Gabbitus\My Documents\Arduino\libraries\WiFi101\src\WiFi.cpp:932:17: error: 'mktime' was not declared in this scope

   t = mktime(&tm);

                 ^

exit status 1
Error compiling for board Arduino/Genuino Mega or Mega 2560.

This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.
```

David

I am using Arduino ide 1.6.12
I have time by michael margolis version 1.50 (latest)
With wifi101 version 0.10.0.
On an arduino mega 2560

The library still does not compile if the sketch includes the <TimeLib.h> (time) library:
I had to force a rebuild between toggling the inclusion of TimeLib.h

The errors returned are 
Arduino: 1.6.12 (Windows 7), Board: "Arduino/Genuino Mega or Mega 2560, ATmega2560 (Mega 2560)"

Build options changed, rebuilding all
C:\Users\laptop\Documents\Arduino\libraries\YiFi101\src\WiFi.cpp: In member function 'uint32_t WiFiClass::getTime()':

C:\Users\laptop\Documents\Arduino\libraries\YiFi101\src\WiFi.cpp:929:13: error: aggregate 'WiFiClass::getTime()::tm tm' has incomplete type and cannot be defined

   struct tm tm;

             ^

C:\Users\laptop\Documents\Arduino\libraries\YiFi101\src\WiFi.cpp:939:17: error: 'mktime' was not declared in this scope

   t = mktime(&tm);

                 ^

exit status 1
Error compiling for board Arduino/Genuino Mega or Mega 2560.

The sketch is from the new ping example
```C++
#include <SPI.h>
#include <WiFi101.h>

#include <TimeLib.h>

char ssid[] = "DRP3";     //  your network SSID (name)
int status = WL_IDLE_STATUS;     // the Wifi radio's status

// Specify IP address or hostname
String hostName = "www.google.com";
int pingResult;
int loopcount = 0;

void setup() {
  // Initialize serial and wait for port to open:
  Serial.begin(115200);
  while (!Serial) {
    ; // wait for serial port to connect. Needed for native USB port only
  }

  // check for the presence of the shield:
  if (WiFi.status() == WL_NO_SHIELD) {
    Serial.println("WiFi shield not present");
    // don't continue:
    while (true);
  }

  // attempt to connect to Wifi network:
  while ( status != WL_CONNECTED) {
    Serial.print("Attempting to connect to WPA SSID: ");
    Serial.println(ssid);
    // Connect to WPA/WPA2 network:
    status =  WiFi.begin(ssid);

    // wait 5 seconds for connection:
    delay(5000);
  }

  // you're connected now, so print out the data:
  Serial.println("You're connected to the network");
  printCurrentNet();
  printWifiData();
}

void loop() {
loopcount++;
  if(loopcount == 10) while(1);
  Serial.print("Pinging ");
  Serial.print(hostName);
  Serial.print(": ");

  pingResult = WiFi.ping(hostName);

  if (pingResult >= 0) {
    Serial.print("SUCCESS! RTT = ");
    Serial.print(pingResult);
    Serial.println(" ms");
  } else {
    Serial.print("FAILED! Error code: ");
    Serial.println(pingResult);
  }

  delay(5000);
}

void printWifiData() {
  // print your WiFi shield's IP address:
  IPAddress ip = WiFi.localIP();
  Serial.print("IP address : ");
  Serial.println(ip);

  Serial.print("Subnet mask: ");
  Serial.println((IPAddress)WiFi.subnetMask());

  Serial.print("Gateway IP : ");
  Serial.println((IPAddress)WiFi.gatewayIP());

  // print your MAC address:
  byte mac[6];
  WiFi.macAddress(mac);
  Serial.print("MAC address: ");
  Serial.print(mac[5], HEX);
  Serial.print(":");
  Serial.print(mac[4], HEX);
  Serial.print(":");
  Serial.print(mac[3], HEX);
  Serial.print(":");
  Serial.print(mac[2], HEX);
  Serial.print(":");
  Serial.print(mac[1], HEX);
  Serial.print(":");
  Serial.println(mac[0], HEX);
  Serial.println();
}

void printCurrentNet() {
  // print the SSID of the network you're attached to:
  Serial.print("SSID: ");
  Serial.println(WiFi.SSID());

  // print the MAC address of the router you're attached to:
  byte bssid[6];
  WiFi.BSSID(bssid);
  Serial.print("BSSID: ");
  Serial.print(bssid[5], HEX);
  Serial.print(":");
  Serial.print(bssid[4], HEX);
  Serial.print(":");
  Serial.print(bssid[3], HEX);
  Serial.print(":");
  Serial.print(bssid[2], HEX);
  Serial.print(":");
  Serial.print(bssid[1], HEX);
  Serial.print(":");
  Serial.println(bssid[0], HEX);

  // print the received signal strength:
  long rssi = WiFi.RSSI();
  Serial.print("signal strength (RSSI): ");
  Serial.println(rssi);

  // print the encryption type:
  byte encryption = WiFi.encryptionType();
  Serial.print("Encryption Type: ");
  Serial.println(encryption, HEX);
  Serial.println();
}
```

It may be worth adding to any ping documentation that a dmz might be required on the yifi101 lan ip, to get the ping to work. 

David

--- a/src/WiFi.cpp
+++ b/src/WiFi.cpp
@@ -17,12 +17,20 @@
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
 
-#if defined(ARDUINO_ARCH_AVR) && (__AVR_LIBC_MAJOR__ < 2)
+#ifdef ARDUINO_ARCH_AVR
+#include <avr/version.h>
+#if (__AVR_LIBC_MAJOR__ < 2)
 #define WIFI_101_NO_TIME_H
+#endif
 #endif
 
 #ifndef WIFI_101_NO_TIME_H
 #include <time.h>
+#endif
+
+#if !defined(_TIME_H_) && !defined(TIME_H)
+// another library overrided the time.h header
+#define WIFI_101_NO_TIME_H
 #endif
 
 #include "WiFi101.h"
@@ -919,6 +927,7 @@ int WiFiClass::ping(IPAddress host, uint8_t ttl)
 uint32_t WiFiClass::getTime()
 {
 #ifdef WIFI_101_NO_TIME_H
+	#warning "No system <time.h> header included, WiFi.getTime() will always return 0"
 	return 0;
 #else
 	tstrSystemTime systemTime;

--- a/src/WiFi.cpp
+++ b/src/WiFi.cpp
@@ -17,7 +17,13 @@
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
 
+#if defined(ARDUINO_ARCH_AVR) && (__AVR_LIBC_MAJOR__ < 2)
+#define WIFI_101_NO_TIME_H
+#endif
+
+#ifndef WIFI_101_NO_TIME_H
 #include <time.h>
+#endif
 
 #include "WiFi101.h"
 
@@ -912,6 +918,9 @@ int WiFiClass::ping(IPAddress host, uint8_t ttl)
 
 uint32_t WiFiClass::getTime()
 {
+#ifdef WIFI_101_NO_TIME_H
+	return 0;
+#else
 	tstrSystemTime systemTime;
 
 	_resolve = (uint32_t)&systemTime;
@@ -946,6 +955,7 @@ uint32_t WiFiClass::getTime()
 	}
 
 	return t;
+#endif
 }
 
 WiFiClass WiFi;

